File: backend/app/services/token_revocation.py
"""Token revocation list — Redis-backed.
When a security event occurs (expired refresh reuse, password reset, session revoke),
the user's access tokens are invalidated by adding the user_id to a revocation set.

Access tokens are short-lived (15 min), so revocation entries auto-expire after 15 min.
"""

from app.config import get_settings
from app.middleware.rate_limit import get_redis

settings = get_settings()
REVOCATION_TTL = settings.access_token_expire_minutes * 60  # Match access token lifetime


async def revoke_user_tokens(user_id: str):
    """Mark all access tokens for a user as revoked.
    They'll be rejected by the auth middleware until they naturally expire.
    """
    redis = await get_redis()
    await redis.set(f"revoked:{user_id}", "1", ex=REVOCATION_TTL)


async def is_user_revoked(user_id: str) -> bool:
    """Check if a user's tokens have been revoked."""
    redis = await get_redis()
    return await redis.exists(f"revoked:{user_id}") > 0

File: backend-rust/src/services/token_revocation.rs
//! Redis-backed token revocation.
//!
//! When a user's tokens need to be invalidated (logout-all, password change,
//! security incident) we set a short-lived key in Redis.  The auth middleware
//! checks this key on every request and rejects tokens issued to revoked users.

use redis::AsyncCommands;

// ── Public API ──────────────────────────────────────────────────────────────

/// Mark all tokens for `user_id` as revoked for `ttl_seconds`.
///
/// The TTL should match (or slightly exceed) the access-token lifetime so that
/// the revocation key naturally expires once all outstanding access tokens have
/// also expired.
pub async fn revoke_user_tokens(
    redis: &redis::Client,
    user_id: &str,
    ttl_seconds: u64,
) -> Result<(), RevocationError> {
    let mut conn = redis
        .get_multiplexed_async_connection()
        .await
        .map_err(|e| RevocationError::Redis(e.to_string()))?;

    let key = revocation_key(user_id);

    conn.set_ex::<_, _, ()>(&key, "1", ttl_seconds)
        .await
        .map_err(|e| RevocationError::Redis(e.to_string()))?;

    tracing::info!(user_id = user_id, ttl = ttl_seconds, "user tokens revoked");

    Ok(())
}

/// Check whether the given user's tokens are currently revoked.
pub async fn is_user_revoked(
    redis: &redis::Client,
    user_id: &str,
) -> Result<bool, RevocationError> {
    let mut conn = redis
        .get_multiplexed_async_connection()
        .await
        .map_err(|e| RevocationError::Redis(e.to_string()))?;

    let key = revocation_key(user_id);

    let exists: bool = conn
        .exists(&key)
        .await
        .map_err(|e| RevocationError::Redis(e.to_string()))?;

    Ok(exists)
}

// ── Helpers ─────────────────────────────────────────────────────────────────

/// Build the Redis key for a user's revocation flag.
fn revocation_key(user_id: &str) -> String {
    format!("revoked:{user_id}")
}

// ── Error type ──────────────────────────────────────────────────────────────

#[derive(Debug, thiserror::Error)]
pub enum RevocationError {
    #[error("redis error: {0}")]
    Redis(String),
}
